src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentCopy = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentCopy = {};
+    } else if (action.type === 'addProperties') {
+      currentCopy = { ...currentCopy, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentCopy = { ...currentCopy };
+
+      for (const key of action.keysToRemove) {
+        delete currentCopy[key];
+      }
+    }
+    stateHistory.push(currentCopy);
```

Comment:
You should push the copy of the object to the `stateHistory` array at the end of the loop cycle but outside of the `if...else if` block to avoid repeating this action in every condition. This adheres to the DRY (Don't Repeat Yourself) principle.